### DISCUSSION TITLE
info on iOS > 14 support of EAP-TTLS authentication method

### DISCUSSION THREAD

**Discussion by s1nnerman89** [strongswan/strongswan#881] [Q&A]
Hi all, sorry if this was common knowledge or already answered somewhere else, but after pouring through the strongSwan wiki, countless Google links and this discussion section, i did not manage to find an answer to my problem.

I am trying to migrate my personal IKEv2 VPN, hosted on my homelab, from using an EAP-TLS based authentication method to an EAP-TTLS based one for added security (or at least what my paranoid self thinks is a safer choice for a personal VPN), but as the title suggests, i'm having problems with my iOS devices. From the strongSwan logs, it seems like the iOS peer doesn't answer the request to initiate an EAP-TTLS authentication, and the iOS default IKEv2 client just fails with a generic "User Authentication Error".

I've tried contacting Apple Support to ask if EAP-TTLS is supported out of the box (which does not appear to be the case, as shown here https://support.apple.com/en-gb/guide/deployment/depae3d361d0/1/web/1.0) or if it can be enabled with Apple Configurator 2; sadly, they didn't know the answer to both these question.

Is what i'm trying to achieve even possible on the default iOS client? For further infos, i'm attaching to this post the logs recorded while trying to establish a connection using EAP-TTLS (personal info have been scrubbed from the file)

[log_strongswan_eap_ttls_iOS.txt](https://github.com/strongswan/strongswan/files/8010969/log_strongswan_eap_ttls_iOS.txt)

UPDATE: after further fiddling with strongSwan's configuration, i managed to get a response from the iOS peer; however, EAP-TTLS does not seem to be supported, as shown by the received EAP_NAK:

`Feb  6 23:23:08 16[ENC] <test-eap-ttls-secrets|1> parsed IKE_AUTH request 2 [ EAP/RES/NAK ]`
`Feb  6 23:23:08 16[IKE] <test-eap-ttls-secrets|1> received EAP_NAK, sending EAP_FAILURE`
`Feb  6 23:23:08 16[ENC] <test-eap-ttls-secrets|1> generating IKE_AUTH response 2 [ EAP/FAIL ]`

The question is then, is there any way to enable support for this method on iOS native client?

**Comment by tobiasbrunner** [comment]
> I am trying to migrate my personal IKEv2 VPN, hosted on my homelab, from using an EAP-TLS based authentication method to an EAP-TTLS based one for added security (or at least what my paranoid self thinks is a safer choice for a personal VPN),

What do you imagine does EAP-TTLS add to the security over EAP-TLS?

> Is what i'm trying to achieve even possible on the default iOS client?

If it doesn't support EAP-TTLS, there is nothing you can do.

> The question is then, is there any way to enable support for this method on iOS native client?

Clearly a question you have to ask Apple.

**Reply by s1nnerman89** [reply]
> What do you imagine does EAP-TTLS add to the security over EAP-TLS?

Well, my idea was to introduce another layer of security to my mobile devices; with EAP-TLS, an "attacker" just needs my phone, then by flipping the VPN switch on he can access my home network. By using EAP-TTLS (in particular with an EAP-MSCHAPv2 authentication protected by the EAP-TLS tunnel) i have an extra layer of security, since now the attacker would also need my password in order to access my home network. Don't know if this makes sense from a security point of view, i'm just an hobbyist who is currently learning the VPN aspect of networking and wanted to add multi-factor authentication to my VPN.

> If it doesn't support EAP-TTLS, there is nothing you can do.

Bummer. Although, after further investigation, it seems that the default iOS client supports PEAP-EAP-MSCHAP. Maybe that's worth a look.

> Clearly a question you have to ask Apple.

I know that, it's just that they gave me unsatisfactory and unclear answers every single time i contacted them; they can't even confirm if the default iOS client supports RFC 4739, so i wanted to ask somebody on the strongSwan side of things for further info, hoping that maybe there was an undocumented way to set up a working implementation of EAP-TTLS with the default iOS VPN client.

**Reply by tobiasbrunner** [reply]
> By using EAP-TTLS (in particular with an EAP-MSCHAPv2 authentication protected by the EAP-TLS tunnel) i have an extra layer of security, since now the attacker would also need my password in order to access my home network.

I see. But that's only the case if EAP-TTLS is used with client certificates, which in many cases it is not (it's mostly used to protect username/password-based EAP methods from passive attackers, e.g. for WPA, which is redundant with IKEv2 as it already provides an encrypted channel). So even if a client supported EAP-TTLS, it might not support configuring a client certificate in addition to username and password (that's the case with iOS, where you can either select a certificate or enter username/password, not both).

> Although, after further investigation, it seems that the default iOS client supports PEAP-EAP-MSCHAP.

Maybe, but it seems unlikely that it supports client authentication for EAP-PEAP.  Definitely not via GUI, as discussed above. But you could try via [configuration profile](https://wiki.strongswan.org/projects/strongswan/wiki/AppleIKEv2Profile) (i.e. set both `PayloadCertificateUUID` and `AuthName`, in addition to `AuthenticationMethod=Certificate` and  `ExtendedAuthEnabled=1`), then configure the server to either request EAP-TTLS or EAP-PEAP with client auth enabled (has to be enabled via `request_peer_auth` for these plugins via [strongswan.conf](https://wiki.strongswan.org/projects/strongswan/wiki/StrongswanConf)).

> if the default iOS client supports RFC 4739

That's most likely not the case. There are not many implementations out there that support it anyway.

**Reply by s1nnerman89** [reply]
Thank you so much for the suggestions, will try them and report my findings on this conversation. For now, i can't make a valid configuration profile with Apple Configurator 2, since everytime i try to install it, it fails with "cannot install profile" and i'm left guessing why.

> request_peer_auth

Interestingly, adding this option to the TTLS plugin breaks a working EAP-TTLS Windows 10 configuration; from the strongswan logs, it feels like the server doesn't even check the client identity. I tried configuring both a EAP-MSCHAPv2 and an EAP-TLS inner method with the win 10 client but the result is always an EAP_FAIL for no apparent reason.

**Comment by s1nnerman89** [comment]
Bumping the thread because i am pleased to say that i managed to establish a working connection using EAP-PEAP-MSCHAPv2 as the authentication method, confirming that iOS clients do indeed support it, albeit not without some tinkering with Apple Configurator 2 following the advice given by Tobias and avoiding using the suggested option `request_peer_auth` on the EAP-PEAP plugin section inside `strongswan.conf`, which breaks the authentication process. More on this later.

No luck for what concerns EAP-TTLS, though. The authentication "handshake" fails with "User Authentication Enabled" on the client and with EAP/FAIL on strongSwan, suggesting that it is not an allowed method.

Anyway, i would like to get, if possible, some insight into the previously mentioned option `request_peer_auth`. From what is reported on the strongSwan wiki,

> charon.plugins.eap-peap.request_peer_auth: Request peer authentication based on a client certificate.

it seems that enabling it allows the peer to authenticate via client certificate. But then, what about the password part of the PEAP-MSCHAPv2 method? Does it make the server ignore the MSCHAPv2 password included into the configuration profile generated by Apple Configurator 2 and makes it expect a client certificate instead? or both?
I'm asking this because, as i said before, enabling it breaks the configuration and i don't know how to troubleshoot it: the connection, which works without the option, fails with a generic "User Authentication Failed" on iOS and with no apparent error on strongSwan. Attached to this message there is an excerpt of the logs captured while attempting a connection with `request_peer_auth` enabled.

[iOS_PEAP_fail_request_peer_auth.txt](https://github.com/strongswan/strongswan/files/8155392/iOS_PEAP_fail_request_peer_auth.txt)

Any help is appreciated. Also, thanks to Tobias for the previous tips, which allowed me to have a working EAP-PEAP-MSCHAPv2 configuration for my mobile devices.

**Reply by tobiasbrunner** [reply]
As I wrote before, enabling that option only works if the client supports client authentication for EAP-PEAP and, as I also mentioned, many do not. They just use the server-authenticated TLS session of that first EAP method (EAP-TTLS is similar) to securely transport the second, username/password-based EAP method (such as EAP-MSCHAPv2 or EAP-MD5). But since IKEv2 already creates a secure channel before such EAP methods are started, this is redundant unless you are using RADIUS and don't trust the connection between VPN gateway and RADIUS server. Such tunneled EAP methods are more useful for insecure channels e.g. to authenticate Wi-Fi clients via WPA. So if the client doesn't support authenticating during the TLS sessions and you don't use RADIUS, there is little benefit in wrapping EAP-MSCHAPv2 in EAP-PEAP/TTLS.

Note that your server log looks weird because EAP-PEAP should not be complete after just an EAP-Identity exchange (and if the client doesn't support client authentication it probably should have been aborted earlier). Have you changed any other options in _charon.plugins.eap-peap_? Could you post more of the log? (Preferably with the log level for _tls_ increased to 2.)
